ocaml/xapi/xapi_vmss.ml

```diff
@@ -0,0 +1,278 @@
+	match ty with
+		| Enum range -> (match (mem value range) with None->err value|Some v->v)
+		| EnumSet range -> (* enumset is a comma-separated string *)
+			let vs = Xstringext.String.split ',' value in
+			List.fold_right 
+				(fun v acc->match (mem v range) with
+					|None->err v
+					|Some v->if acc="" then v else (v^","^acc)
+				)
+				vs ""
+		| IntRange (min,max) ->
+			let v=try int_of_string value with _->err value in
+			if (v<min || v>max) then err value else value
+		| ReqValue required_value -> if value <> required_value then err value else value
+		|String -> value
+
+let with_ks ~kss ~fn =
+	let field,kss=kss in
+	let corrected_values = List.filter (fun cv->cv<>None) (List.map (fun ks-> fn field ks) kss) in
+	if List.length corrected_values < 1 then []
+	else (match List.hd corrected_values with None->[]|Some cv->cv)
+
+let assert_req_values ~field ~ks ~vs =
+	(* each required values in this ks must match the one in the vs map this key/value belongs to*)
+	let req_values = List.fold_right
+		(fun (k,attr) acc->match attr with(ReqValue rv),_->(k,rv)::acc|_->acc) ks []
+	in
+	(if vs<>[] then
+		List.iter (fun (k,rv)->
+			if (List.mem_assoc k vs) then (if rv<>(List.assoc k vs) then err field k rv)
+		) req_values
+	)
+
+let merge xs ys = (* uses xs elements to overwrite ys elements *)
+	let nys = List.map (fun (ky,vy)->if List.mem_assoc ky xs then (ky,(List.assoc ky xs)) else (ky,vy)) ys in
+	let nxs = List.filter (fun (kx,_)->not(List.mem_assoc kx nys)) xs in
+	nxs@nys
+
+let assert_key ~field ~ks ~key ~value =
+	debug "assert_key: field=%s key=[%s] value=[%s]" field key value;
+	(* check if the key and value conform to this ks *)
+	(if not (List.mem_assoc key ks)
+		then err field key value
+	else
+		assert_value ~field ~key ~attr:(List.assoc key ks) ~value
+	)
+
+let assert_keys ~ty ~ks ~value ~db =
+	let value = merge value db in
+	with_ks ~kss:ks ~fn:
+		(fun field (xt,ks) ->
+			debug "assert_keys: field=%s xt=[%s] ty=[%s]" field xt ty;
+			if (xt=ty) then Some
+			(
+				assert_req_values ~field ~ks ~vs:value;
+				(* for this ks, each key value must be valid *)
+				List.map (fun (k,v)-> k,(assert_key ~field ~ks ~key:k ~value:v)) value
+			)
+			else None
+		)
+
+let assert_all_keys ~ty ~ks ~value ~db =
+	let value = merge value db in
+	with_ks ~kss:ks ~fn:
+		(fun field (xt,ks)->
+			debug "assert_all_keys: field=%s xt=[%s] ty=[%s]" field xt ty;
+			if (xt=ty) then Some
+			(
+				assert_req_values ~field ~ks ~vs:value;
+				(* add missing keys with default values *)
+				let value = List.map (fun (k,(kt,default))->if List.mem_assoc k value then (k,(List.assoc k value)) else (k,default)) ks in
+				
+				(* remove extra unexpected keys *)
+				let value = List.fold_right (fun (k,v) acc->if List.mem_assoc k ks then (k,v)::acc else acc) value [] in
+				(* for this ks, each key value must be valid *)
+				List.map (fun (k,v)-> k,(assert_key ~field ~ks ~key:k ~value:v)) value
+			)
+			else None
+		)
```

Comment:
There are already some functions like this for handling key-value pair lists in map_check.ml - that might be a good place to add them.